Remove unreachable lines of sampling code

MarcoGorelli opened the pull request:
I may be wrong here, but I think these lines are unreachable. I was trying to write a sensible test which covers them, but I think that might not be possible.

The function `instantiate_steppers` is only ever called from `assign_step_methods`.

In `assign_step_methods`, there is
```python
    selected_steps = defaultdict(list)
    for var in model.free_RVs:
        if var not in assigned_vars:
            # determine if a gradient can be computed
            has_gradient = var.dtype not in discrete_types
            if has_gradient:
                try:
                    tg.grad(model.logpt, var)
                except (AttributeError, NotImplementedError, tg.NullTypeGradError):
                    has_gradient = False
            # select the best method
            selected = max(
                methods,
                key=lambda method, var=var, has_gradient=has_gradient: method._competence(
                    var, has_gradient
                ),
            )
            selected_steps[selected].append(var)
```

We have two cases:

- If `model` has no unnassigned free random variables, then `selected_steps` will be empty and so the loop `    for step_class, vars in selected_steps.items():` in `instantiate_steppers` will never be reached.

- if `model` has unnasigned free random variables, then `selected_steps[selected]` will be a list with `var` in it, and so `if len(vars) == 0` will never be True.

In both cases, the `continue` in
```python
        if len(vars) == 0:
            continue
```
won't be reached.

----

cc @michaelosthege - does this look right to you, or am I missing something?

MarcoGorelli committed:
remove unreachable code

michaelosthege reviewed (changes requested):
The docstring should be improved and type annotations could be added:
```
steps : step function or vector of step functions
        One or more step functions that have been assigned to some subset of
        the model's parameters. Defaults to None (no assigned variables).
```
In contrast to what the docstring says, the code assumes that `steps` is a list and it _can_ be empty.

This code should trigger line 138:
```
with pymc3.Model() as model:
    n = pymc3.Normal("n")
    u = pymc3.Normal("u")

    steppers = pymc3.sampling.instantiate_steppers(
        _model=model,
        steps=[
            pymc3.Slice([u])
        ],
        selected_steps={
            pymc3.Metropolis : [n],
            pymc3.NUTS : [],
        }
    )
    
    print(steppers)
```

Granted, it is not typically used that way, but instead of removing that `len(vars)` check, I would suggest to indent lines 134-138 and invert the condition.

MarcoGorelli committed:
clarify docs, invert guard

MarcoGorelli committed:
clarify that steps is a sequence

MarcoGorelli commented on pymc3/sampling.py:
```diff
@@ -111,31 +111,31 @@ def instantiate_steppers(_model, steps, selected_steps, step_kwargs=None):
+    steps : sequence of step functions
         One or more step functions that have been assigned to some subset of
-        the model's parameters. Defaults to None (no assigned variables).
+        the model's parameters.
```
`None` has no `.append` method

michaelosthege replied:
SHould be like this:

```
steps : list
        A list of zero or more step function instances that have been assigned to some subset of
        the model's parameters.
```

MarcoGorelli commented on pymc3/sampling.py:
```diff
@@ -111,31 +111,31 @@ def instantiate_steppers(_model, steps, selected_steps, step_kwargs=None):
     -------
-    methods : list
-        List of step methods associated with the model's variables.
+    methods : list or step
+        List of step methods associated with the model's variables, or step method
+        if there is only one.
```
if it's a list of length 1, then just `steps[0]` is returned

MarcoGorelli commented:
Thanks for your review and help here, I've updated - for now I've just updated the docstring and inverted the condition, as I haven't figured out how to type some of the objects here yet (e.g. variables)

michaelosthege commented on pymc3/sampling.py:
```diff
@@ -111,31 +111,31 @@ def instantiate_steppers(_model, steps, selected_steps, step_kwargs=None):
+        the model's parameters.
     selected_steps : dictionary of step methods and variables
-        The step methods and the variables that have were assigned to them.
+        The step methods and the (possibly zero) variables that have were assigned to them.
```
And here, it should say
```
selected_steps : dict
    A dictionary that maps a step method class to a list of model variables.
```

michaelosthege commented on pymc3/sampling.py:
```diff
@@ -111,31 +111,31 @@ def instantiate_steppers(_model, steps, selected_steps, step_kwargs=None):
+        the model's parameters.
     selected_steps : dictionary of step methods and variables
-        The step methods and the variables that have were assigned to them.
+        The step methods and the (possibly zero) variables that have were assigned to them.
```
And here, it should say
```
selected_steps : dict
    A dictionary that maps a step method class to a list of model variables.
```

MarcoGorelli replied:
Sure - I've added `zero or more` here because it's the fact that there could be zero _here_ which was causing the `continue` line to be uncovered

MarcoGorelli committed:
:memo: update docstring

twiecki commented:
Thanks @MarcoGorelli and @michaelosthege.
